tsl/src/continuous_aggs/refresh.c

```diff
@@ -374,3 +374,54 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 			 "continuous aggregate \"%s\" is already up-to-date",
 			 NameStr(cagg->data.user_view_name));
 }
+
+/*
+ * Refresh all continuous aggregates on a hypertable.
+ *
+ * The refreshing happens in a single transaction. For this to work correctly,
+ * there must be no new invalidations written in the refreshed region during
+ * the refresh. Therefore, the caller is responsible for proper locking to
+ * ensure there are no invalidations.
+ */
+void
+continuous_agg_refresh_all(const Hypertable *ht, int64 start, int64 end)
+{
+	Catalog *catalog = ts_catalog_get();
+	List *caggs = ts_continuous_aggs_find_by_raw_table_id(ht->fd.id);
+	Dimension *dim = hyperspace_get_open_dimension(ht->space, 0);
+	InternalTimeRange refresh_window = {
+		.type = ts_dimension_get_partition_type(dim),
+		.start = start,
+		.end = end,
+	};
+	ListCell *lc;
+
+	LockRelationOid(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+					AccessExclusiveLock);
+	continuous_agg_invalidation_threshold_set(ht->fd.id, refresh_window.end);
```

Comment:
This is not strictly true if a previous call to refresh had an `end` value that was later. Not sure if that could be an issue, but it does look strange. Please check.

Reply:
Not sure what you mean. FWIW, the function only updates the threshold if new value is greater than the old one. This is old code, so haven't changed.

Reply:
In that case there should be no problems.